Syntactic function arity parsing

ncik-roberts opened the pull request:
This is the parsing half of syntactic function arity. Unlike the upcoming typechecking/translation PR, this PR does not change the semantics of any existing OCaml program. But it does place new requirements on ppxes, so we should wait to merge it until some internal PRs for ppxes land.

@goldfirere for review please.

This PR transplants bits of the upstream PR (ocaml/ocaml#12236) to flambda-backend.

At a high level:
- The OCaml parser now keeps track of function arity, encoding this in the parsetree using the Jane Syntax framework.
- The sanity checks that run on ASTs produced by ppxes now require that functions were encoded using the Jane Syntax framework.
- For now, typechecking drops this arity information. (It constructs another parsetree that doesn't have any arity information and typechecks that.) 

### How to review
- Review the first two commits (`Add some locals tests` and `Make Jane Syntax treat attributes as inner-to-outer`) in isolation. See "Things of note".
- Then look at the entire diff of the PR.

This PR does not share much code with the upstream PR, except for some pieces of the parser.

### Things of note

#### Making Jane Syntax treat attributes as inner-to-outer

One of the early commits ensures the following things hold:
  1. Jane Syntax should interpret later attributes before earlier attributes. ("Later" and "earlier" are with respect to the order of attributes in, say, the `pexp_attributes` list.)
  2. When the user writes something like `[: :] [@attr]`, the `[@attr]` attribute should be placed later in the list than any Jane Syntax attribute.

(1.) did not hold previously, but (2.) did. This PR makes both properties hold. It's important for (1.) to hold because, without it, this PR would be incompatible with the way that ppxlib merges attributes. For example, suppose a ppx replaces `[%ppx_extension]` with `fun y z -> y + z`, which is encoded in Jane Syntax as `(fun[@jane.erasable._builtin] y -> fun[@jane.erasable._builtin.param] z -> (y + z)[@jane.erasable._builtin.end])`. Then,

```ocaml
(* source program *)
(fun x -> [%ppx_extension])

(* encoding as Jane Syntax parsetree when handed to the ppx *)
(fun[@jane.erasable._builtin] x -> [%ppx_extension] [@jane.erasable._builtin.end])

(* result of running ppx, keeping in mind that ppxlib merges attributes by placing outer attributes after
   inner ones *)
(fun[@jane.erasable._builtin] x -> fun[@jane.erasable._builtin][@jane.erasable._builtin.end] y -> fun[@jane.erasable._builtin.param] z -> (y + z)[@jane.erasable._builtin.end])
```

So, for `fun[@jane.erasable._builtin][@jane.erasable._builtin.end] y -> ...`, Jane Syntax likewise needs to interpret `@jane.erasable._builtin.end` before `@jane.erasable._builtin`, because otherwise it will think the AST is invalid.

#### `alloc_mode` on function type constraints

The parsetree encoded in Jane Syntax is different: there's an `alloc_mode` field of the type constraint. This is because the Jane Street OCaml parser treats this construct specially:

```ocaml
let local_ f x : int -> int -> int = fun y z -> x + y + z
```
We want this type to be treated by the type-checker as `local_ (int -> int -> int)`. The old parser accomplished this by putting an attribute on the `Pexp_constraint` node, but given that we're no longer constructing a `Pexp_constraint` node directly, I instead add a new field to the `type_constraint`. I add some tests in this PR because we weren't testing this behavior before.

### Merge this as-is or wait for typechecking?

TODO @ncik-roberts : I need to check that typedtree is unchanged if we want to merge this.

The main reason I can think to not merge this is that it constructs, and then throws away, a dummy parsetree just for purposes  of typechecking. You'd expect this to impact minor heap allocations, and it looks like it does. (For a relevant benchmark, this is 1.9% worse: 15.9 TB vs. 15.6 TB.) But minor heap allocations are cheap, so maybe we want to merge this anyway in the interest of merging smaller, simpler PRs.

goldfirere reviewed (commented):
Instead of accumulating lots and lots of comments in a complete review, I've decided it's best just to send comments every day they accumulate. More to come. (Note to self: stopped in `Parser.all_params_as_newtypes`.)

ncik-roberts commented on ocaml/parsing/jane_syntax.ml:
```diff
@@ -284,29 +302,305 @@ module Immutable_arrays = struct
+    | Some (End_expression_body, expr) ->
+        let params = List.rev rev_params in
+        Some (params, function_constraint, Pfunction_body expr)
+    | None ->
```
We cannot, as this pattern shadows `expr`.

Ekdohibs replied:
(Note: in the future this could be done with the pattern matchings changes + allowing `when` under or-patterns: `Some (End_expression_body, expr) | (None when expr match expr) -> ...`)

ncik-roberts commented on ocaml/parsing/ast_invariants.ml:
```diff
@@ -71,8 +74,24 @@ let iterator =
+    match body with
+    | Pfunction_cases _ -> ()
+    | Pfunction_body _ ->
+        match
```
Yes, and I couldn't tell you why I didn't write this in the first place.

ncik-roberts commented on ocaml/parsing/jane_syntax.mli:
```diff
@@ -78,12 +76,59 @@ module Immutable_arrays : sig
+    | Pconstraint of Parsetree.core_type
+    | Pcoerce of Parsetree.core_type option * Parsetree.core_type
+
+  type alloc_mode = Local | Global
```
I see the temptation, but we don't actually benefit from any code sharing in this patch so I move to not do this. (Another nudge: I like being able to document the constructors of this type here. This certainly doesn't rule out `alloc_mode` being an alias to another type, but it is a small nudge.)

ncik-roberts commented on ocaml/parsing/jane_syntax.mli:
```diff
@@ -78,12 +76,59 @@ module Immutable_arrays : sig
+        [x] carries the location of the identifier, whereas [loc] is
+        the location of the [(type x)] as a whole.
+
+        Multiple parameters [(type a b c)] are represented as multiple
```
I answered upstream and put the same comment here. If you would like to revise that comment, let's discuss here: https://github.com/ocaml/ocaml/pull/12236#discussion_r1261471806

ncik-roberts commented on ocaml/parsing/jane_syntax.mli:
```diff
@@ -78,12 +76,59 @@ module Immutable_arrays : sig
+module N_ary_functions : sig
+  type function_body =
+    | Pfunction_body of Parsetree.expression
+    | Pfunction_cases of Parsetree.case list * Location.t * Parsetree.attributes
```
I've now correctly transplanted the missing comment from the upstream patch.

ncik-roberts commented on ocaml/parsing/jane_syntax.ml:
```diff
@@ -284,29 +302,305 @@ module Immutable_arrays = struct
+        Desugaring_error.raise expr Missing_closing_embedding
+
+  (* Returns remaining unconsumed attributes on outermost expression *)
+  let of_expr =
```
Indeed, that was a notable omission. How does this BNF look?

ncik-roberts commented on ocaml/parsing/jane_syntax_parsing.ml:
```diff
@@ -506,10 +506,11 @@ let parse_embedding_exn ~loc ~payload ~name ~embedding_syntax =
 let find_and_remove_jane_syntax_attribute =
-  let rec loop rest ~rev_prefix =
-    match rest with
+  (* Recurs on [rev_prefix] *)
```
My impression is that it both needed to change and _did_ change. (See the below comment "See Note [Outer attributes at end] in jane_syntax.ml" on `make_jane_syntax`.) Or are you talking about something different?

goldfirere replied:
Yes -- I must have just missed it. Looks good.

ncik-roberts commented on ocaml/parsing/jane_syntax.ml:
```diff
@@ -284,29 +302,305 @@ module Immutable_arrays = struct
+    | None -> None
+    | Some (ext_name, attributes) -> begin
+        let expr = { expr with pexp_attributes = attributes } in
+        match Jane_syntax_parsing.Embedded_name.components ext_name with
```
Sounds good.

ncik-roberts commented on ocaml/parsing/parser.mly:
```diff
@@ -705,6 +708,70 @@ let class_of_let_bindings ~loc lbs body =
     mkclass ~loc (Pcl_let (lbs.lbs_rec, List.rev bindings, body))
 
+let all_params_as_newtypes =
+  let exception Found_val in
```
I agree with your disentanglement. I spent some time thinking about a nicer way to present this and arrived at this alternate implementation. If you like this more, I'll do it in the upstream patch, too.

goldfirere replied:
Not as good as Haskell, but good enough. :)

ncik-roberts commented on ocaml/parsing/jane_syntax.ml:
```diff
@@ -284,29 +302,305 @@ module Immutable_arrays = struct
+    let raise expr err = raise (Error (expr.pexp_loc, err))
+  end
+
+  let expand_n_ary_expr expr =
```
I've taken a shot at moving things into `Attribute_node_with_payload`. I think it feels somewhat better than the previous approach -- the main tension I feel is that the `of_expr` function now has to do some case analysis to figure out what attributes weren't consumed by Jane Syntax. But the old approach was just complex in a different way in this regard, so I think on the whole I like your suggestion better.

goldfirere commented on ocaml/parsing/parser.mly:
```diff
@@ -2437,17 +2535,39 @@ class_type_declarations:
-  | expr        %prec below_SEMI  { $1 }
-  | expr SEMI                     { $1 }
-  | mkexp(expr SEMI seq_expr
+%inline or_function(EXPR):
```
Did not review parser changes carefully; assuming that the grammar changes are adapted from upstream.

ncik-roberts replied:
That sounds right to me.

goldfirere commented on ocaml/parsing/pprintast.ml:
```diff
@@ -687,9 +694,15 @@ and sugar_expr ctxt f e =
   | _ -> false
 
-and expression ctxt f x =
+(* Jane Syntax defies the property "if an expression has attributes,
```
I don't quite follow the plan here. It looks like `jane_syntax_expr` looks to see if there are attributes after peeling off the jane-syntax attributes, so isn't that sufficient?

goldfirere commented on ocaml/parsing/pprintast.ml:
```diff
@@ -1908,6 +1933,76 @@ and unboxed_constant _ctxt f (x : Jane_syntax.Unboxed_constants.t)
   | Float (x, suffix) -> pp f "#%a" constant (Pconst_float (x, suffix))
   | Integer (x, suffix) -> pp f "#%a" constant (Pconst_integer (x, Some suffix))
 
+and function_param ctxt f (x : Jane_syntax.N_ary_functions.function_param) =
```
I assume this is all the same as upstream.

ncik-roberts replied:
That's right.

goldfirere commented on ocaml/testsuite/tests/parsetree/locations_test.compilers.reference:
```diff
@@ -266,12 +266,18 @@ Ptop_def
             expression (//toplevel//[2,1+4]..[2,1+35])
               Pexp_constraint
               expression (//toplevel//[2,1+25]..[2,1+35])
+                attribute "jane.erasable._builtin.param"
```
All the additions here make me worry about performance cost here of the more elaborate encoding.

ncik-roberts replied:
Per [this comment](https://github.com/ocaml-flambda/flambda-backend/pull/1548#issuecomment-1662282983), the attribute encoding is now less heavy and this test's output is updated as such.

goldfirere commented on ocaml/testsuite/tests/typing-local/local.ml:
```diff
@@ -2569,3 +2569,67 @@ Error: This local value escapes its region
         Adding 1 more argument will make the value non-local
 |}]
+
+(* Escaping uncurried functions *)
```
Assuming these are the same as upstream. Did not review carefully.

ncik-roberts replied:
These tests are new: I added them in the first commit because I noticed there were no tests where a function was constrained at the local mode.

I don't add the upstream tests in this PR because we haven't yet changed the way that arity works, so there's nothing new to test.

ncik-roberts replied:
Richard and I looked at these tests together and we think they look fine, so I'll resolve this comment.

goldfirere commented on ocaml/typing/typeclass.ml:
```diff
@@ -274,9 +274,11 @@ let type_constraint val_env sty sty' loc =
 let make_method loc cl_num expr =
   let open Ast_helper in
   let mkid s = mkloc s loc in
-  Exp.fun_ ~loc:expr.pexp_loc Nolabel None
```
I almost wonder whether `Ast_helper.Exp.fun_` should get a new implementation in terms of the new syntax. Do you think that would cause trouble? It would seem to offer some future-proofing.

ncik-roberts replied:
Jane Syntax uses `Ast_helper.Exp.fun_`, so I think it would cause trouble. I went a different route by putting an alert on `Exp.fun_` and `Exp.function_` so it's easy to audit places we've used them and so the alert text can redirect people to Jane Syntax.

ncik-roberts commented:
Just now, I unintentionally pushed to the wrong remote, so my latest force-push backs that out. I haven't responded to the new comments yet.

ncik-roberts commented on ocaml/parsing/ast_invariants.ml:
```diff
@@ -107,6 +125,7 @@ let iterator =
     | Pexp_new id -> simple_longident id
     | Pexp_record (fields, _) ->
       List.iter (fun (id, _) -> simple_longident id) fields
+    | Pexp_function _ | Pexp_fun _ -> non_jane_syntax_function loc
```
I think it's a good idea to try to get this PR merged before all ppxes are migrated to the new system. That means I will need to:
  (a) remove this check for ast_invariants
  (b) ensure that existing ppxes build with this PR.

I'm leaving this comment unresolved as a proxy for both changes, though it's only attached to (a).

ncik-roberts commented:
TL;DR: The rebase was messy, and the commit history is much less useful as a result. I recommend just reviewing the overall diff.

The rebase is messy because #1417 introduced the need to support things like:

```ocaml
let f x y (type a : immediate) z = ...
```

We need to embed this information in the n-ary functions encoding, as the `(type a : immediate)` shouldn't interrupt a function's arity. There are lots of intermediate commits in this PR, and indeed many that change the attribute encoding, and I didn't want to support `(type a : immediate)` at every step of the way. Instead of fixing up every intermediate commit to get it compiling and working, I just fixed up one of the latest commits.

In other words, most of the early commits no longer even compile, let alone work.

Here is a brief summary of the changes I made in order to support `(type a : immediate)`:
  * I moved the chunk of code that encodes/decodes layout annotations as payloads so I could reuse it for n-ary functions.
  * I embedded the layout annotation as a `_builtin` attribute, but I checked that layouts were enabled (at any maturity) before allowing the `(type a : immediate)` syntax. The usual layout-maturity check is run to ensure that `immediate` is only allowed at alpha.

ncik-roberts commented:
The most recent pair of commits transplants a commit from upstream that embeds the parameter location in the parsetree. This isn't strictly needed for this PR, but it will simplify the next PR.

ncik-roberts commented:
> TODO @ncik-roberts : I need to check that typedtree is unchanged if we want to merge this.

I ran a test against the stdlib and observed that the produced typedtree is unchanged, with one exception: the location of a function is now marked as ghost in the typedtree for constructions like `let f (type t) x = x`. I have reason to believe this change is fine: the function node in `let f x = x` is marked as ghost already in the upstream compiler, and merlin does not have any trouble with this. (Furthermore, I think it's just correct for the `Texp_function` in `let f (type t) x = x` to be marked as ghost &mdash; it doesn't appear literally in the source text.)

goldfirere commented:
OK. It sounds like this is ready for re-review from me, given the rebase on top of #1417. Is that correct?

ncik-roberts commented:
Yes, this is now ready for re-review from you.

goldfirere commented on ocaml/testsuite/tests/formatting/test_locations.dlocations.ocamlc.reference:
```diff
@@ -6,6 +6,8 @@
         pattern (test_locations.ml[17,534+8]..[17,534+11])
           Ppat_var "fib" (test_locations.ml[17,534+8]..[17,534+11])
         expression (test_locations.ml[17,534+14]..[19,572+34])
+          attribute "jane.erasable._builtin"
```
I don't see this attribute listed in the BNF in `Jane_syntax`. I think it probably should, because it's part of the little grammar here. But upon closer inspection, I don't know how this gets here at all. The empty suffix seems to arise only from `Top_level`, but that seems never to happen (except by reading an existing `jane.erasable._builtin`). So something is off somewhere (quite possibly my understanding).

To be clear, I didn't review the code in Jane_syntax all that closely -- it will be short-lived and most mistakes would be easily caught during routine testing.

ncik-roberts replied:
This is the top-level attribute that introduces the Jane Syntax construct. (Same as, say, `jane.non_erasable.comprehension`.) My decision to not include it in the BNF mirrors Antal's decision to not include `jane.non_erasable.comprehension` in the `Comprehension` BNF.

This attribute is inserted by `Jane_syntax_parsing`, not by the syntactic arity&ndash;related code in `Jane_syntax.

goldfirere replied:
Ah -- that makes sense. Thanks.

goldfirere commented on ocaml/typing/typecore.ml:
```diff
@@ -196,16 +196,16 @@ type error =
 exception Error_forward of Location.error
 
-let error_of_filter_arrow_failure ~explanation in_function ty_fun
+let error_of_filter_arrow_failure ~explanation ~first ty_fun
```
I do not anticipate interesting changes here from the previous review; did not review. Please let me know if there's a spot I should worry about.

ncik-roberts replied:
That seems right to me.

ncik-roberts commented on ocaml/parsing/jane_syntax.ml:
```diff
@@ -111,13 +129,180 @@ end
+      (string Location.loc * layout_annotation option) list
+  end
+end = struct
+  module Desugaring_error = struct
```
Good idea, done.

ncik-roberts commented:
My latest rebase is on top of the Uniqueness PR, which does affect n-ary lambdas because of constructs like `let local_ f x : int -> int = ...`. (We need the type annotation to be interpreted at the provided mode.) Prior to the rebase, the only mode allowed in that position was `local_`, but now `unique` and `once` can be provided there, too.

goldfirere commented:
This is all good from my point of view. Thanks!

ncik-roberts commented:
The last rebase resolves some trivial conflicts with #1761. I'll go ahead and merge this after some internal testing finishes.
